### ISSUE TITLE
Select all and then un-check one of the checkbox result in wrong checkbox unticked

### ISSUE THREAD

**Issue by vthai** [jbetancur/react-data-table-component#346]
## Describe the bug
After select all items in the table(by checking the select all checkbox), proceed to uncheck one item results in wrong item being un-ticked.

## To Reproduce
1. Load data from server to RDT
2. select all by clicking on "select-all" check box
3. pick the second item to uncheck
4. the last item being unchecked

## Expected behavior
The second item should be unchecked instead of the last one

## Versions
- RDT 4.0.0

## Additional context
I suspect this is very much to do with the [`removeItem`](https://github.com/jbetancur/react-data-table-component/blob/d3741261e07de28d7ec851a6a46eb87eddc5b2ab/src/DataTable/util.js#L44) method in `util.js`, 
 as the method employs Javascript object comparison:

`newArray.findIndex(a => a === item)`

Which is not robust at all as that is based on object reference comparison, so

```
let a = { "name": "A" };
let b = { "name": "A" };
console.log(a === b); // false

```
This in turn results in the wrong removal of item in data set

`newArray.splice(newArray.findIndex(a => a === item), 1);`

As `findIndex` would return -1 in such case, resulted in the last item being removed)

As we cannot guarantee that object in the data set are always the same reference, we should do the `keyField` comparison instead, maybe something like:

```
export const removeItem = (array, item, keyField  =  'id') => {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a[keyField] === item[keyField]), 1);

  return newArray;
};
```

**Comment by jbetancur** [comment]
You are correct about the compare. The compare only works if it’s the same reference. I actually fixed this here with selecta but looks like I missed remove item https://github.com/jbetancur/react-data-table-component/blob/d3741261e07de28d7ec851a6a46eb87eddc5b2ab/src/DataTable/util.js#L94

**Comment by jbetancur** [comment]
I’ll get a patch out today on this

**Comment by jbetancur** [comment]
What’s strange about this is I can’t dupe it but I know it’s wrong.

That being said, I don’t think that code has changed since version 1.0. How are you setting your data? Does each data record have a unique id?

**Comment by jbetancur** [comment]
This is now fixed in 4.0.3

**Comment by vthai** [comment]
> What’s strange about this is I can’t dupe it but I know it’s wrong.
> 
> That being said, I don’t think that code has changed since version 1.0. Does each data record have a unique id? Are there duplicates?

Yeah, it is hard to reproduce actually, as the data are sent from server axios might have their way on how to reuse or create JavaScript object, the data item do have the `id` field.

**Comment by vthai** [comment]
> This is now fixed in 4.0.3

Thanks a lot for the super quick fix! Have a great weekend.
